Support multiple PODs in GPC PCDs

ax0 opened the pull request:
This PR adds support for multiple PODs in GPC PCDs.

Summary of changes:
- Reformulate `GPCArgs` as an intersection type including a record of named POD PCDs, which are assumed to be prefixed with `podpcd_`.
- Appropriate changes to the examples including a test in `gpc-pcd`.

artwyman commented on packages/pcd/gpc-pcd/src/GPCPCD.ts:
```diff
@@ -97,7 +97,15 @@ export type GPCPCDArgs = {
   id?: StringArgument;
   // TODO(POD-P3): Support PODValue of multiple types.
-};
+} /**
```
I don't think the documentation here will be properly extracted for developer docs (see https://docs.pcd.team).  Docs need to be attached to something named, like a type or a field.
An option might be to make a named type for this Record to put detailed docs there.  The GPCArgs type docs should also explain the presence of these special fields.

artwyman commented on packages/pcd/gpc-pcd/src/GPCPCD.ts:
```diff
@@ -97,7 +97,15 @@ export type GPCPCDArgs = {
-};
+} /**
+ * POD objects to prove about. Each object is identified by name and prefixed
+ * with {@link PODPCD_ARG_PREFIX} followed by an underscore.  These are not
```
I think for external developers we should say `podpcd` and expect them to use that prefix as-is rather than the constant (you could link to the constant in a parenthetical so devs know it's available).  I think we're unlikely to feel we can safely change this prefix once it's in use (and we have any compatibility guarantees), so given that I'd suggest erring on the side of simplicity and clarity rather than maximizing generality.

artwyman commented on packages/pcd/gpc-pcd/src/GPCPCD.ts:
```diff
@@ -97,7 +97,15 @@ export type GPCPCDArgs = {
+ *
+ * See {@link GPCProofConfig} and {@link GPCRevealedClaims} for more
+ * information.
+ */ & Record<PODPCDArgName, PCDArgument<PODPCD>>;
```
This is clever.  I woudln't have expected something like this to work, but I assume since you modified consumer-client you've tested it.

I'm not sure I have enough depth of context on TypeScript/Javascript and how the PCD framework is used to know if there could be any gotchas with this approach.  We should get input from @ichub before finalizing this.

artwyman commented on packages/pcd/gpc-pcd/src/GPCPCDPackage.ts:
```diff
@@ -62,13 +65,48 @@ async function checkProofArgs(args: GPCPCDArgs): Promise<{
   const proofConfig = deserializeGPCProofConfig(args.proofConfig.value);
 
-  if (!args.pod.value) {
+  // Filter out PODPCD records.
```
This is tricky and clever.
Maybe too clever, since it took me a while to puzzle over what it's doing.
What do you think of an alternative like this (may not compile, but you can hopefully get the gist:
```ts
    const podRecord = Object.fromEntries(Object.entries(args).filter(([n, _v]) => n.startsWith(PODPCD_ARG_PREFIX)));
```

ax0 replied:
That is reasonable, though I was thinking it would be good to catch invalid arguments and throw as I do further down, but perhaps that is unnecessary. I'll make the change.

artwyman replied:
Ah, fair point.  Catching invalid arguments is a reasonable thing to do, though mixing it with extracting the podpcd arguments seems a bit odd.  In the case of PCD requests we'll certainly notice missing args, but not unrecognized ones.  The nature of how TypeScript types work does tend to encourage ignoring unknown fields rather than being strict about them.  I didn't put any unknown field checking in the GPC types (like GPCProofConfig) though I wonder whether I should at some point.  It depends how we approach compatibility, I suppose.  I.e. if we add a new field, do we want clients which don't know about it to reject the config or treat it like an old version?

artwyman commented on packages/pcd/gpc-pcd/src/GPCPCDPackage.ts:
```diff
@@ -62,13 +65,48 @@ async function checkProofArgs(args: GPCPCDArgs): Promise<{
-  if (!isPODPCD(podPCD)) {
-    throw new Error("Wrong PCD type provided for PODPCD");
-  }
+
```
My review isn't complete yet.  Need to pick up from here...

artwyman reviewed (commented):
Haven't finished reviewing yet, but I need to run. Submitting my comments on the overall approach in case you're online and have time to start thinking about it.  Will finish reviewing later.

artwyman commented on packages/pcd/gpc-pcd/src/GPCPCDPackage.ts:
```diff
@@ -62,13 +65,42 @@ async function checkProofArgs(args: GPCPCDArgs): Promise<{
+  const pods = Object.fromEntries(
+    await Promise.all(
+      Object.entries(podRecord).map(async ([prefixedPODName, podPCDArg]) => {
+        const podNameMatch = prefixedPODName.match(PODPCD_ARG_REGEXP);
```
I think it would be better to check the prefix separately, then check the PODName using POD_NAME_REGEX (or checkPODName).  Otherwise PODPCD_ARG_REGEX becomes something we could easily miss changing if/when POD_NAME_REGEX changes.

artwyman commented on packages/pcd/gpc-pcd/src/GPCPCDPackage.ts:
```diff
@@ -62,13 +65,42 @@ async function checkProofArgs(args: GPCPCDArgs): Promise<{
+
+        const podPCD = await PODPCDPackage.deserialize(podPCDArg.value.pcd);
+        if (!isPODPCD(podPCD)) {
+          throw new Error("Wrong PCD type provided for PODPCD");
```
Include the arg name here

artwyman commented on packages/pcd/gpc-pcd/src/GPCPCDPackage.ts:
```diff
@@ -251,7 +283,7 @@ export function getProveDisplayOptions(): ProveDisplayOptions<GPCPCDArgs> {
         defaultVisible: true
       },
-      pod: {
+      [`${PODPCD_ARG_PREFIX}_pod0`]: {
```
This isn't going to be sufficient if the args have different names.  What do you see if you try to prove here using a name other than podpcd_pod0?  The prove screen needs this information in order to give the user the right interface to select a PCD, and the validate function below is what filters them down to PODs.  Making the prove screen deal with argument names in a generic way is going to need more work.

artwyman commented on packages/pcd/gpc-pcd/test/gpc-pcd.spec.ts:
```diff
@@ -66,15 +72,26 @@ describe("GPCPCD should work", async function () {
 
-    const pod = POD.sign(sampleEntries, privateKey);
-    const podPCD = new PODPCD(uuid(), pod);
+    const pod0 = POD.sign(sampleEntries0, privateKey);
```
I'd suggest switching to a name other than pod0 everywhere, to ensure the system is no longer relying on that hardcoded name.

artwyman commented on examples/pod-gpc-example/src/gpcExample.ts:
```diff
@@ -419,7 +419,7 @@ export async function gpcDemo(): Promise<boolean> {
       value: serializeGPCProofConfig(pcdProofConfig)
     },
-    pod: {
+    [`${PODPCD_ARG_PREFIX}_pod0`]: {
```
Use `podpcd_pod0` for clarity here.  I don't think we should make external devs use the prefix constant.

Also this is a tutorial, so I think the example should now include multiple PODs, and names friendlier than "pod0" so we can demonstrate how that works.  You could rework this section to use a similar multi-item proof to the one used for the non-PCD case above, though it can have a simpler config.  The goal of this section is to demonstrate what's unique to the GPCPCD layer, not all the possible features of GPCs.

artwyman commented on apps/consumer-client/src/pages/examples/gpc-proof.tsx:
```diff
@@ -195,7 +200,7 @@ export function openGPCPopup(
       userProvided: false
     },
-    pod: {
+    [`${PODPCD_ARG_PREFIX}_pod0`]: {
```
There needs to be some way of exercising multiple PODs from consumer-client to validate this, particularly the UI aspects.

artwyman commented on apps/consumer-client/src/pages/examples/add-pcd.tsx:
```diff
@@ -637,7 +637,7 @@ async function addGPCPCD(
       value: gpcConfig
     },
-    pod: {
+    [`${PODPCD_ARG_PREFIX}_pod0`]: {
```
There needs to be some way of exercising multiple PODs from consumer-client to validate this, particularly the UI aspects.

artwyman reviewed (changes requested):
The types and raw arg handling look good here, but I think there's more work needed for the proving UI to properly support these flexible args when their name isn't `podpcd_pod0`.  I think some way for the UI to cleanly let the user select multiple PODs is a requirement here.  There may be some optional polish around how to let the requestor set user-friendly names and descriptions on each POD for the UI too.

I'm actually not familiar with the proving screen code, and would have to start reading it to figure out what's possible.  Rob may be a good resource to talk to for guidance on how that works and ideas about how this idea might integrate with it, so I suggest reaching out to him.

ichub commented on packages/pcd/gpc-pcd/src/GPCPCD.ts:
```diff
@@ -20,6 +21,23 @@ export type {
+ * Regular expression for `PODPCDArgName` values.
+ */
+export const PODPCD_ARG_REGEXP = new RegExp(
+  String.raw`(${PODPCD_ARG_PREFIX})_([A-Za-z_]\w*)$`
```
```suggestion
  String.raw`(${PODPCD_ARG_PREFIX})_([A-Za-z_]\w)*$`
```
is this right?

ichub commented on packages/pcd/gpc-pcd/src/GPCPCDPackage.ts:
```diff
@@ -62,13 +65,42 @@ async function checkProofArgs(args: GPCPCDArgs): Promise<{
+          throw new Error(`No PODPCD value provided for POD ${podName}`);
+        }
+
+        const podPCD = await PODPCDPackage.deserialize(podPCDArg.value.pcd);
```
this may or may not throw depending on user input and deserialization logic

artwyman replied:
I think that's okay.  prove() should fail if we're given an invalid PCD.  Would you suggest something else?

ichub replied:
One consideration I'm thinking about is UX. For example, I wonder what is shown on the prove screen given an exception at this location. Errors bubble up to either a Developer, User, or Rollbar, so I generally like to think about how they will manifest to one of those audiences. For example, when an error occurs during the loading of a Podbox pipeline, it's recorded into an array of errors that then get saved for viewing later. Just expanding on my thought process here.

ichub commented on packages/pcd/gpc-pcd/src/GPCPCD.ts:
```diff
@@ -47,17 +77,6 @@ export type GPCPCDArgs = {
-   * See {@link GPCProofConfig} and {@link GPCRevealedClaims} for more
-   * information.
-   */
-  pod: PCDArgument<PODPCD>;
```
i suggest turning `pod` into `pods: PODPCDRecordArg` so as to not pollute this namespace, rather than what you do below with `& PODPCDRecordArg`

artwyman replied:
That's an interesting idea.  Cleaner at this level, though it'll presumably require some more change to the logic of the prove screen in order to display one "arg" to the user as multiple, each with its own description, validator, etc, since we want each POD to be selectable separately.
Beyond this use case in particular, it would make sense to have a general abstraction for multiple args by name in a group like this, and/or args in an array.  If those existed they'd apply here easily, but they might be more complicated to build from scratch.

ichub commented on packages/pcd/gpc-pcd/src/GPCPCDPackage.ts:
```diff
@@ -62,13 +65,42 @@ async function checkProofArgs(args: GPCPCDArgs): Promise<{
-    throw new Error("Wrong PCD type provided for PODPCD");
-  }
+
+  const pods = Object.fromEntries(
```
```suggestion
  const pods: sometype = Object.fromEntries(
```

adding a type annotation here would help readability

ichub reviewed (commented):
i think this approach makes sense overall - i'd be curious to see how exactly it integrates with the prove screen for cases when a pod needs to be provided by the user before fully committing to it though.

ax0 commented:
I've incorporated @ichub's suggestion of implementing/using a record argument type encapsulating records with (primitive, i.e. non-record) arguments as values. I opted for non-recursive records for simplicity and modified the validator and display logic accordingly: Record arguments are flattened out and args are taken from the record values, then from the records, then from the default args, with validator parameters being combined (with the same order of precedence). I've updated all of the examples and put in basic entry-based POD filtering to show that it all works.

artwyman commented on apps/consumer-client/src/pages/examples/add-pcd.tsx:
```diff
@@ -144,30 +147,43 @@ export default function Page(): JSX.Element {
         </button>
         <br />
+        <br />
+        POD content to sign: <br />
```
Zupass card POD example to sign:

artwyman commented on apps/consumer-client/src/pages/examples/add-pcd.tsx:
```diff
@@ -144,30 +147,43 @@ export default function Page(): JSX.Element {
-        <button
-          onClick={() => {
-            setPodContent(EXAMPLE_POD_CONTENT);
+        <textarea
```
I unfortunately made some visual changes here in #1765 which are going to cause you conflicts.  Should be pretty easy to resolve so long as you know what I intended to change:
- Put the POD and GPC stuff in a separate ExampleContainer section.
- Change spelling of function names from "Pod" to "POD".
- Move the add POD button up above the GPC config (which you also did)
- Add a different "Test GPCs" folder config

artwyman commented on apps/consumer-client/src/pages/examples/add-pcd.tsx:
```diff
@@ -637,9 +656,18 @@ async function addGPCPCD(
-      argumentType: ArgumentTypeName.PCD
+    pods: {
+      value: {
+        podZero: {
```
`examplePOD` maybe?  Not sure why `podZero` would be a meaningful name if it's not for counting 0, 1, 2.

artwyman commented on apps/consumer-client/src/podExampleConstants.ts:
```diff
@@ -43,18 +43,63 @@ export const EXAMPLE_GPC_CONFIG = `{
+    },
+    "ticketPOD": {
+      "entries": {
+        "ticketID": {
```
Where would I get a POD to satisfy this test case?  It would be helpful for quick testing if the examples on the add-pcd page would meet this requirement somehow, so it wouldn't take manually typing POD values to set up a test.

artwyman commented on apps/consumer-client/src/pages/examples/gpc-proof.tsx:
```diff
@@ -195,13 +195,26 @@ export function openGPCPopup(
+        },
+        ticketPOD: {
+          argumentType: ArgumentTypeName.PCD,
+          pcdType: PODPCDPackage.name,
```
It would be awesome if this could be a PODTicketPCD and "just work".
Seems a relatively easy extension, but doesn't need to be part of this PR.

artwyman commented on packages/pcd/gpc-pcd/src/GPCPCD.ts:
```diff
@@ -33,7 +39,17 @@ export type GPCPCDInitArgs = {
+/**
+ * Defines the essential parameters required for creating a {@link GPCPCD}. This
+ * includes some fixed arguments as well as a variable number of arguments with
+ * the `podpcd_` prefix defined by the {@link PODPCDRecordArg} type.
```
podpcd_ prefix is no longer a thing

artwyman commented on packages/pcd/gpc-pcd/src/GPCPCDPackage.ts:
```diff
@@ -62,12 +62,32 @@ async function checkProofArgs(args: GPCPCDArgs): Promise<{
+  const pods: Record<PODName, POD> = Object.fromEntries(
+    await Promise.all(
+      Object.entries(args.pods.value).map(async ([podName, podPCDArg]) => {
+        if (podName.match(POD_NAME_REGEX) === null) {
```
Use checkPODName function?

artwyman commented on packages/pcd/gpc-pcd/src/GPCPCD.ts:
```diff
@@ -33,7 +39,17 @@ export type GPCPCDInitArgs = {
 
 /**
- * Defines the essential parameters required for creating a {@link GPCPCD}.
+ * Record argument type encapsulating a record of POD PCD arguments.
```
Discuss that the names need to correspond to the names used in the proof config.

artwyman commented on packages/pcd/gpc-pcd/src/GPCPCDPackage.ts:
```diff
@@ -251,23 +271,42 @@ export function getProveDisplayOptions(): ProveDisplayOptions<GPCPCDArgs> {
+            }
+
+            if (params?.proofConfig !== undefined) {
+              const proofConfig = deserializeGPCProofConfig(params.proofConfig);
```
This can throw.  I think it may be best not to throw from the validator, but simply return false, with a customized notFoundMessage

artwyman commented on packages/lib/pcd-types/src/pcd.ts:
```diff
@@ -369,10 +392,17 @@ export type RawValueType<T extends Argument<any, unknown>> =
+  T extends RecordArgument<infer S, infer U>
+    ? (
+        s: S
+      ) => (value: RawValueType<U>, params: U["validatorParams"]) => boolean
```
Could `s` (name) become an extra parameter here before value, rather than needing the extra layer of function call?

artwyman commented on apps/passport-client/components/shared/PCDArgs.tsx:
```diff
@@ -56,20 +58,49 @@ export function PCDArgs<T extends PCDPackage>({
+  // Flatten record arguments (if any), keeping track of the parent argument to
+  // properly mutate (cf. `setArg`) as well as inheriting argument fields from
+  // the record argument.  Validator parameters are also combined.
+  const flattenedArgs = Object.entries(args).flatMap(([argName, arg]) => {
```
Add a type here for clarity, so a reader knows what flattenedArgs is meant to contain.

artwyman commented on apps/passport-client/components/shared/PCDArgs.tsx:
```diff
@@ -56,20 +58,49 @@ export function PCDArgs<T extends PCDPackage>({
   options?: ArgsDisplayOptions<ArgsOf<T>>;
 }): JSX.Element {
   const [showAll, setShowAll] = useState(false);
+
```
Ivan should review this file, being more familiar with the React UI layer

artwyman reviewed (changes requested):
Love this!  I'm glad that flattening the record worked out, and that you were able to get validation to work.

I've made some suggestions in comments.
@ichub should also please take a look, particularly at pcd.ts and PCDArgs.tsx since I don't have full context on what's most important there.

artwyman commented on packages/pcd/gpc-pcd/src/GPCPCD.ts:
```diff
@@ -39,17 +39,16 @@ export type GPCPCDInitArgs = {
  */
 export type PODPCDRecordArg = RecordArgument<
   PODName,
   PCDArgument<PODPCD, { notFoundMessage?: string; proofConfig?: string }>
```
I didn't notice this before.  It seems wrong that proofConfig is here rather than on the RecordArgument, which is where you're specifying it in the example.  This controls the validatorParams?  Does it apply to both?  I didn't ever focus on the validatorParams being a specified type.  Maybe make this a named type so you can use it in both places, and also in the args of the validator function?

artwyman commented on packages/pcd/gpc-pcd/src/GPCPCDPackage.ts:
```diff
@@ -274,39 +272,49 @@ export function getProveDisplayOptions(): ProveDisplayOptions<GPCPCDArgs> {
-          (podName) =>
-          (podPCD, params): boolean => {
-            if (podPCD.type !== PODPCDTypeName) {
+        validate: (podName, podPCD, params): boolean => {
```
Give these args types?

artwyman commented on packages/pcd/gpc-pcd/src/GPCPCDPackage.ts:
```diff
@@ -274,39 +272,49 @@ export function getProveDisplayOptions(): ProveDisplayOptions<GPCPCDArgs> {
+          }
+
+          if (params?.proofConfig !== undefined) {
+            const proofConfig = ((): GPCProofConfig | undefined => {
```
I understand the use of the lambda here, but what's its value?
I think if the try/catch was directly in this block the `return false` could be part of the catch.
The only thing which would have to change is to make the `proofConfig` declaration a let rather than a const so it can be set inside the try.

ax0 replied:
Sorry, force of habit. I tend to avoid the pattern `let x; x = ...;`. I'll make the change!

artwyman commented on packages/pcd/gpc-pcd/src/GPCPCDPackage.ts:
```diff
@@ -274,39 +272,49 @@ export function getProveDisplayOptions(): ProveDisplayOptions<GPCPCDArgs> {
+                if (e instanceof TypeError) {
+                  params.notFoundMessage = e.message;
+                }
+              }
```
If this isn't a TypeError, it's falling through to default-return undefined.  That's pretty non-obvious.  I'd recommend an explict return, or `throw e;` instead.

artwyman reviewed (approved):
Some stylistic suggestions but they're pretty trivial.  Otherwise this looks great!

Please hold off for Ivan to review before merging, though.

artwyman commented:
FYI, also tried all the new Consumer Client cases in my own environment and they worked.

ichub commented on apps/consumer-client/src/pages/examples/add-pcd.tsx:
```diff
@@ -172,6 +162,14 @@ export default function Page(): JSX.Element {
         />
         <br />
+        <button
+          onClick={(): Promise<void> =>
```
```suggestion
          onClick={() =>
```

im pretty sure you don't need this

ichub commented on apps/consumer-client/src/pages/examples/add-pcd.tsx:
```diff
@@ -184,14 +182,44 @@ export default function Page(): JSX.Element {
+        />
+        <br />
         <button
           onClick={(): Promise<void> =>
```
```suggestion
          onClick={() =>
```

ichub commented on apps/passport-client/components/shared/PCDArgs.tsx:
```diff
@@ -56,20 +58,66 @@ export function PCDArgs<T extends PCDPackage>({
   const [showAll, setShowAll] = useState(false);
+
+  // Flatten record arguments (if any), keeping track of the parent argument to
+  // properly mutate (cf. `setArg`) as well as inheriting argument fields from
```
what is `cf.`?

ax0 replied:
Confer/compare.

ichub commented on apps/passport-client/components/shared/PCDArgs.tsx:
```diff
@@ -56,20 +58,66 @@ export function PCDArgs<T extends PCDPackage>({
+  // Flatten record arguments (if any), keeping track of the parent argument to
+  // properly mutate (cf. `setArg`) as well as inheriting argument fields from
+  // the record argument.  Validator parameters are also combined.
+  type flattenedArgTriple = [
+    string | undefined,
+    string,
+    Argument<PrimitiveArgumentTypeName>
+  ];
```
generally prefer to declare types outside of functions (i.e. in module scope), and for their names to start with capital letters.

ichub commented on apps/passport-client/components/shared/PCDArgs.tsx:
```diff
@@ -104,48 +153,75 @@ export function PCDArgs<T extends PCDPackage>({
-          ...args[argName],
-          value
-        }
+        ...(parentArgName !== undefined
+          ? {
+              [parentArgName]: {
+                ...args[parentArgName],
+                value: {
+                  ...args[parentArgName].value,
+                  [argName]: {
+                    ...args[parentArgName].value[argName],
+                    value
+                  }
+                }
+              }
+            }
+          : {
+              [argName]: {
+                ...args[argName],
+                value
+              }
+            })
```
this is quite confusing and i anticipate will be difficult to maintain unless we break it up somehow

ichub commented on apps/passport-client/components/shared/PCDArgs.tsx:
to make sure this works properly, i suggest doing some manual fuzzing and trying out as many weird combinations as possible.

ichub commented on packages/lib/pcd-types/src/pcd.ts:
```diff
@@ -256,10 +256,33 @@ export enum ArgumentTypeName {
   Object = "Object",
   StringArray = "StringArray",
   PCD = "PCD",
+  Record = "Record",
```
Is the idea with record that it's like like `Object` but in a way that's compatible with the `POD` data type? I wonder if the name `Record` could be something else instead that communicates the intent somehow. E.g. I feel like it probably makes sense to use the word `POD` here somehow.

artwyman replied:
It's not POD-specific.  I think a record contains multiple named PCDArgs (each with their own type, value, and parameters, and separate serialization rules) vs. an Object which is a single PCDArg which is a JSON-stringifiable object as its value.  The name Record makes some sense in Typescript, but I'm open to clearer suggestions.  Map?  NamedArgs?

ichub reviewed (approved):
nice
